[Materials] Add dark mode support

pxlcoder opened the pull request:
#### 0e26af9fa9a5720aa1be96a47aade5e82361e12e
<pre>
[Materials] Add dark mode support
<a href="https://bugs.webkit.org/show_bug.cgi?id=288152">https://bugs.webkit.org/show_bug.cgi?id=288152</a>
<a href="https://rdar.apple.com/145091546">rdar://145091546</a>

Reviewed by Abrar Rahman Protyasha.

Materials and vibrancy should adapt to dark mode. This patch implements that
functionality for CoreMaterial backed visual effects, as well as SwiftUI-based
effects.

The color scheme of an effect is determined as part of creating
`AppleVisualEffectData`, and then applied by `RemoteLayerTreePropertyApplier`.

* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-chrome-dark-expected.txt: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-chrome-dark.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-chrome-expected.txt:
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-dark-expected.txt: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-dark.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-expected.txt:
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-thick-dark-expected.txt: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-thick-dark.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-thick-expected.txt:
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-thin-dark-expected.txt: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-thin-dark.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-thin-expected.txt:
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-ultra-thin-dark-expected.txt: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-ultra-thin-dark.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-ultra-thin-expected.txt:
* LayoutTests/apple-visual-effects/apple-visual-effect-dark-mode-expected-mismatch.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-dark-mode.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-vibrancy-effects-dark-expected.txt: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-vibrancy-effects-dark.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-vibrancy-effects-expected.txt:
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-chrome-dark-expected.txt: Added.
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-chrome-expected.txt:
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-dark-expected.txt: Added.
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-expected.txt:
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-thick-dark-expected.txt: Added.
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-thick-expected.txt:
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-thin-dark-expected.txt: Added.
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-thin-expected.txt:
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-ultra-thin-dark-expected.txt: Added.
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-ultra-thin-expected.txt:
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-vibrancy-effects-dark-expected.txt: Added.
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-vibrancy-effects-expected.txt:
* Source/WebCore/PAL/pal/cocoa/CoreMaterialSoftLink.h:

Add dark mode material constants.

* Source/WebCore/PAL/pal/cocoa/CoreMaterialSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/CoreMaterialSPI.h:
* Source/WebCore/platform/cocoa/AppleVisualEffect.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/cocoa/AppleVisualEffect.h:

Introduce a `colorScheme` field to `AppleVisualEffectData`. This will be used to
select the appropriate material on the UI-side.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAppleVisualEffect):

Use the used color scheme of the renderer to determine the color scheme of the
visual effect.

* Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.h:
* Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.swift:
(MaterialHostingView.body):

`materialEffect` uses the environment of the parent view. For this reason, the
environment modifier is applied to an `AnyView` which wraps the view with the
material effect applied. The environment&apos;s color scheme is set to achieve the
desired appearance.

(WKMaterialHostingSupport.updateHostingLayer(_:materialEffectType:colorScheme:cornerRadius:)):
(WKMaterialHostingSupport.updateHostingView(_:contentView:materialEffectType:colorScheme:cornerRadius:)):
(WKMaterialHostingSupport.updateHostingLayer(_:materialEffectType:cornerRadius:)): Deleted.
(WKMaterialHostingSupport.updateHostingView(_:contentView:materialEffectType:cornerRadius:)): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::materialRecipeForAppleVisualEffect):
(WebKit::hostedMaterialColorSchemeForAppleVisualEffectData):
(WebKit::applyVisualStylingToLayer):
(WebKit::updateAppleVisualEffect):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
(-[WKMaterialHostingView updateMaterialEffectType:colorScheme:cornerRadius:]):
(-[WKMaterialHostingView updateMaterialEffectType:cornerRadius:]): Deleted.

Canonical link: <a href="https://commits.webkit.org/290803@main">https://commits.webkit.org/290803@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd00d4c5ac4d7646a8ee815c72da317f7ada436c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91090 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10633 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/125 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96099 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41860 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93140 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11042 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18953 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70005 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27532 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94091 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8417 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82541 "Found 1 new API test failure: TestWebKitAPI.WKWebExtensionAPIDevTools.PortMessagePassingToBackground (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50331 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8188 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/106 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40990 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78510 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/123 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98078 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18297 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79015 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18556 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78365 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78215 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22725 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/78 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14392 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18298 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23630 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18024 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21485 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19803 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.swift:
```diff
@@ -91,17 +92,19 @@ private struct MaterialHostingView<P: MaterialHostingProvider>: View {
 
         if let effect = MaterialHostingView<P>.resolvedMaterialEffect(for: materialEffectType) {
-            view.materialEffect(effect, in: .rect(cornerRadius: cornerRadius))
+            AnyView(view.materialEffect(effect, in: .rect(cornerRadius: cornerRadius)))
```
what happens if you get rid of `AnyView` and instead use https://developer.apple.com/documentation/swiftui/view/preferredcolorscheme(_:) instead of the color scheme environment?

pxlcoder replied:
Doesn't work either.

rr-codes replied:
😭

pxlcoder committed:
[Materials] Add dark mode support
https://bugs.webkit.org/show_bug.cgi?id=288152
rdar://145091546

Reviewed by Abrar Rahman Protyasha.

Materials and vibrancy should adapt to dark mode. This patch implements that
functionality for CoreMaterial backed visual effects, as well as SwiftUI-based
effects.

The color scheme of an effect is determined as part of creating
`AppleVisualEffectData`, and then applied by `RemoteLayerTreePropertyApplier`.

* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-chrome-dark-expected.txt: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-chrome-dark.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-chrome-expected.txt:
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-dark-expected.txt: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-dark.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-expected.txt:
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-thick-dark-expected.txt: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-thick-dark.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-thick-expected.txt:
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-thin-dark-expected.txt: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-thin-dark.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-thin-expected.txt:
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-ultra-thin-dark-expected.txt: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-ultra-thin-dark.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-blur-material-ultra-thin-expected.txt:
* LayoutTests/apple-visual-effects/apple-visual-effect-dark-mode-expected-mismatch.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-dark-mode.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-vibrancy-effects-dark-expected.txt: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-vibrancy-effects-dark.html: Added.
* LayoutTests/apple-visual-effects/apple-visual-effect-vibrancy-effects-expected.txt:
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-chrome-dark-expected.txt: Added.
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-chrome-expected.txt:
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-dark-expected.txt: Added.
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-expected.txt:
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-thick-dark-expected.txt: Added.
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-thick-expected.txt:
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-thin-dark-expected.txt: Added.
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-thin-expected.txt:
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-ultra-thin-dark-expected.txt: Added.
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-blur-material-ultra-thin-expected.txt:
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-vibrancy-effects-dark-expected.txt: Added.
* LayoutTests/platform/ios/apple-visual-effects/apple-visual-effect-vibrancy-effects-expected.txt:
* Source/WebCore/PAL/pal/cocoa/CoreMaterialSoftLink.h:

Add dark mode material constants.

* Source/WebCore/PAL/pal/cocoa/CoreMaterialSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/CoreMaterialSPI.h:
* Source/WebCore/platform/cocoa/AppleVisualEffect.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/cocoa/AppleVisualEffect.h:

Introduce a `colorScheme` field to `AppleVisualEffectData`. This will be used to
select the appropriate material on the UI-side.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAppleVisualEffect):

Use the used color scheme of the renderer to determine the color scheme of the
visual effect.

* Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.h:
* Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.swift:
(MaterialHostingView.body):

`materialEffect` uses the environment of the parent view. For this reason, the
environment modifier is applied to an `AnyView` which wraps the view with the
material effect applied. The environment's color scheme is set to achieve the
desired appearance.

(WKMaterialHostingSupport.updateHostingLayer(_:materialEffectType:colorScheme:cornerRadius:)):
(WKMaterialHostingSupport.updateHostingView(_:contentView:materialEffectType:colorScheme:cornerRadius:)):
(WKMaterialHostingSupport.updateHostingLayer(_:materialEffectType:cornerRadius:)): Deleted.
(WKMaterialHostingSupport.updateHostingView(_:contentView:materialEffectType:cornerRadius:)): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::materialRecipeForAppleVisualEffect):
(WebKit::hostedMaterialColorSchemeForAppleVisualEffectData):
(WebKit::applyVisualStylingToLayer):
(WebKit::updateAppleVisualEffect):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
(-[WKMaterialHostingView updateMaterialEffectType:colorScheme:cornerRadius:]):
(-[WKMaterialHostingView updateMaterialEffectType:cornerRadius:]): Deleted.

Canonical link: https://commits.webkit.org/290803@main
